orttraining/orttraining/core/session/training_session.cc

```diff
@@ -193,8 +197,30 @@ Status TrainingSession::ConfigureForTraining(
         weight_names_to_train, mixed_precision_config.use_fp16_initializers, fp32_weight_name_to_fp16_node_arg));
   }
 
-  if (config.use_pipeline) {
-    ORT_RETURN_IF_ERROR(InsertPipelineOps());
+  if (config.pipeline_config.has_value()) {
+    TrainingConfigurationResult::PipelineConfigurationResult pipeline_result{};
+    ORT_RETURN_IF_ERROR(InsertPipelineOps(pipeline_result.forward_waited_event_name,
+                                          pipeline_result.forward_recorded_event_name,
+                                          pipeline_result.backward_waited_event_name,
+                                          pipeline_result.backward_recorded_event_name,
+                                          pipeline_result.forward_waited_output_name,
+                                          pipeline_result.forward_recorded_output_name,
+                                          pipeline_result.backward_waited_output_name,
+                                          pipeline_result.backward_recorded_output_name));
+    // The following loop is for not to fetch tensors not in this pipeline stage.
+    for (size_t i = 0; i < config.pipeline_config.value().fetch_names.size(); ++i) {
+      auto name = config.pipeline_config.value().fetch_names[i];
+      const auto* node_arg = model_->MainGraph().GetNodeArg(name);
+      if (!node_arg) {
+        // This pipelie stage doesn't contain this name.
+        // Let's not to fetch it.
+        continue;
+      }
+      pipeline_result.fetch_names.push_back(name);
+    }
+    pipeline_result.pipeline_stage_id = config.distributed_config.world_rank / 
+      (config.distributed_config.data_parallel_size * config.distributed_config.horizontal_parallel_size);
```

Comment:
could you add some comments about the assignment? Do we want to wrap it in a more general distributed schedule func/class?

Reply:
Like `DistributedContext.GetPipelineParallelId()`, `DistributedContext.GetDataParallelId()`, `DistributedContext.GetHorizonParallelId()`? Conceptually, the three IDs forms a 3-D tensor and each tensor element is a GPU.

Reply:
Looks like DistributedContext needs to be largely revised. I just created another task for that.

Reply:
Sure, we can do that in a different PR.

---
In reply to: [420984032](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r420984032) [](ancestors = 420984032)